core/dbt/cli/flags.py

```diff
@@ -29,29 +34,47 @@ def assign_params(ctx):
                 if hasattr(self, param_name):
                     raise Exception(f"Duplicate flag names found in click command: {param_name}")
                 object.__setattr__(self, param_name.upper(), param_value)
+                if ctx.get_parameter_source(param_name) == ParameterSource.DEFAULT:
+                    params_assigned_from_default.add(param_name)
             if ctx.parent:
-                assign_params(ctx.parent)
+                assign_params(ctx.parent, params_assigned_from_default)
 
-        assign_params(ctx)
+        params_assigned_from_default = set()  # type: Set[str]
+        assign_params(ctx, params_assigned_from_default)
 
         # Get the invoked command flags
         if hasattr(ctx, "invoked_subcommand") and ctx.invoked_subcommand is not None:
             invoked_subcommand = getattr(import_module("dbt.cli.main"), ctx.invoked_subcommand)
             invoked_subcommand.allow_extra_args = True
             invoked_subcommand.ignore_unknown_options = True
             invoked_subcommand_ctx = invoked_subcommand.make_context(None, sys.argv)
-            assign_params(invoked_subcommand_ctx)
+            assign_params(invoked_subcommand_ctx, params_assigned_from_default)
+
+        if not user_config:
+            profiles_dir = getattr(self, "PROFILES_DIR", None)
+            user_config = read_user_config(profiles_dir) if profiles_dir else None
+
+        # Overwrite default assignments with user config if available
+        if user_config:
+            for param_assigned_from_default in params_assigned_from_default:
+                user_config_param_value = getattr(user_config, param_assigned_from_default, None)
+                if user_config_param_value is not None:
+                    object.__setattr__(
+                        self, param_assigned_from_default.upper(), user_config_param_value
+                    )
 
         # Hard coded flags
-        object.__setattr__(self, "WHICH", ctx.info_name)
+        object.__setattr__(
+            self, "WHICH", ctx.protected_args[0] if ctx.protected_args else ctx.info_name
```

Comment:
I'm not sure I follow what this is doing... According to [the docs](https://click.palletsprojects.com/en/8.1.x/api/#click.Context.protected_args) `protected_args` is used for certain extra parsing scenarios and that doesn't sound like that we're looking for when using `flags.WHICH`?

Reply:
Here I was looking to get an accurate value for `WHICH` regardless of whether the context was based on the CLI group or a subcommand. I assume we'll want callers going through the CLI group context because it does a lot of setup (and even implements some CLI functionality, such as `--version`).

In either case the caller would be responsible for figuring out what to set `name` to when constructing a `Context` object programmatically, which is then used to set `flags.VERSION`:
```python
>>> from dbt.cli.main import cli
>>> from dbt.cli.flags import Flags
# Caller would need to infer that "run" is the right name to set based on args.
>>> ctx = cli.make_context(cli.name, ["--partial-parse", "run"])
>>> flags = Flags(ctx)
>>> flags.WHICH
'cli'
```

Perhaps `invoked_subcommand` is what we should use here, if available. However `invoked_subcommand` is only available if the command is actually being _invoked_, as opposed to based on the provided `args`. So it would be `None` in the example above. 

I went with `protected_args` because of the docs hinting that it is used to implement nested parsing, but it does feel fragile as well.

Reply:
Related note - Currently on feature/click-cli, flags.WHICH is `dbt` [at this point](https://github.com/dbt-labs/dbt-core/compare/feature/click-cli#diff-43832707bfec8cdaa49c13f004d9b3296585827f7e68ef07fba7ad1d5e6eb7c6R62), not the subcommand name when called directly. So it's technically not accurate unless we [recreate flags further along in the subcommand method](https://github.com/dbt-labs/dbt-core/compare/feature/click-cli#diff-43832707bfec8cdaa49c13f004d9b3296585827f7e68ef07fba7ad1d5e6eb7c6R231), which I understand we're looking to move away from doing. Using `invoked_subcommand` when available should at least fix the issue from the CLI interface.

Reply:
As far as I'm aware, the real function of `flags.WHICH` is to support conditional checks in user code for `{{ if flags.WHICH = ... }}`. That is, conditional behavior that only wants to run during certain commands, and not others.

links:
- the `flags` module is exposed to the context [here](https://github.com/dbt-labs/dbt-core/blob/c3be975783e065c4dc42c10fca29aca8e5584945/core/dbt/context/base.py#L626-L642)
- https://docs.getdbt.com/reference/dbt-jinja-functions/flags
- [functional test showing this behavior](https://github.com/dbt-labs/dbt-core/blob/main/tests/functional/schema_tests/test_schema_v2_tests.py)

Reply:
Oh I'm glad we had this conversation!  Yes `invoked_subcommand` is exactly what we should be using-- In my head that's what `info_name` did-- I probably C/P'd the wrong line in the original code (`info_name` is right above `invoked_subcommand` in the output of `pp vars(flags)`).

Reply:
Updated to `invoked_subcommand` if available, falling back on ctx.info_name. I think it'd still be worth looking into `protected_args` more but I'll punt that to a follow-up if that's alright; feels a bit out of scope for this work.